test_pokemon.py

```diff
@@ -25,14 +27,73 @@ def test_relu(self):
         x = report_observation(observation)
         action = pkmn.choose_action(x, bookkeeper)
         # Judgement and shadow claw are garbage; so is switching
         assert(action == "move 1" or action == "move 3")
+    
+    def test_preprocessing(self):
+        # First test: give it a plausible start state. Note that the format of the input to preprocess_observation
+        # is not finalised. If I can get the simulator to return a JSON object of the state, we might be looking
+        # at two preprocess_observation classes; one for fighting people online (where you can't just muck around
+        # in the guts of the game and get it to return what you want) and one for training.
+        state_changes, indices = preprocess_observation('{"active":[{"moves":[{"move":"Thunderbolt","id":"thunderbolt","pp":24,"maxpp":24,"target":"normal","disabled":false},{"move":"Shadow Claw","id":"shadowclaw","pp":24,"maxpp":24,"target":"normal","disabled":false},{"move":"Aerial Ace","id":"aerialace","pp":32,"maxpp":32,"target":"any","disabled":false},{"move":"Judgment","id":"judgment","pp":16,"maxpp":16,"target":"normal","disabled":false}]}],"side":{"name":"Bob","id":"p2","pokemon":[{"ident":"p2: Arceus","details":"Arceus-Fighting","condition":"402/402","active":true,"stats":{"atk":245,"def":297,"spa":297,"spd":297,"spe":297},"moves":["thunderbolt","shadowclaw","aerialace","judgment"],"baseAbility":"multitype","item":"fistplate","pokeball":"pokeball","ability":"multitype"},{"ident":"p2: Aggron","details":"Aggron, L10, M","condition":"39/39","active":false,"stats":{"atk":32,"def":46,"spa":22,"spd":22,"spe":20},"moves":["roar","heavyslam","rockslide","toxic"],"baseAbility":"sturdy","item":"aggronite","pokeball":"pokeball","ability":"sturdy"},{"ident":"p2: Dragonite","details":"Dragonite, L10, F","condition":"43/43","active":false,"stats":{"atk":37,"def":29,"spa":30,"spd":30,"spe":26},"moves":["dragonclaw","firepunch","roost","earthquake"],"baseAbility":"multiscale","item":"lumberry","pokeball":"pokeball","ability":"multiscale"},{"ident":"p2: Uxie","details":"Uxie, L10","condition":"40/40","active":false,"stats":{"atk":25,"def":36,"spa":25,"spd":36,"spe":29},"moves":["psyshock","yawn","stealthrock","psychic"],"baseAbility":"levitate","item":"leftovers","pokeball":"pokeball","ability":"levitate"},{"ident":"p2: Cacturne","details":"Cacturne, L10, F","condition":"39/39","active":false,"stats":{"atk":33,"def":22,"spa":33,"spd":22,"spe":21},"moves":["swordsdance","seedbomb","suckerpunch","spikes"],"baseAbility":"waterabsorb","item":"leftovers","pokeball":"pokeball","ability":"waterabsorb"},{"ident":"p2: Druddigon","details":"Druddigon, L10, F","condition":"40/40","active":false,"stats":{"atk":34,"def":28,"spa":22,"spd":28,"spe":19},"moves":["dragontail","suckerpunch","gunkshot","aerialace"],"baseAbility":"sheerforce","item":"lifeorb","pokeball":"pokeball","ability":"sheerforce"}]},"State":["40/40","272/272","46/46","42/42","37/37","41/41",1]}\nDEADBEEF\n')
```

Comment:
This'll be easier to understand if you create a helper function for preprocess_observation that just takes the dictionary as its input. Then you test as follows:

        obs = {
            "State": [
                "40/40",
                "272/272",
                "46/46",
                "42/42",
                "37/37",
                "41/41",
                1
            ],
            "active": [
                {
                    "moves": [
                        {
                            "disabled": False,
                            "id": "thunderbolt",
                            "maxpp": 24,
                            "move": "Thunderbolt",
                            "pp": 24,
                            "target": "normal"
                        },
                        {
                            "disabled": False,
                            "id": "shadowclaw",
                            "maxpp": 24,
                            "move": "Shadow Claw",
                            "pp": 24,
                            "target": "normal"
                        },
                        {
                            "disabled": False,
                            "id": "aerialace",
                            "maxpp": 32,
                            "move": "Aerial Ace",
                            "pp": 32,
                            "target": "any"
                        },
                        {
                            "disabled": False,
                            "id": "judgment",
                            "maxpp": 16,
                            "move": "Judgment",
                            "pp": 16,
                            "target": "normal"
                        }
                    ]
                }
            ],
            "side": {
                "id": "p2",
                "name": "Bob",
                "pokemon": [
                    {
                        "ability": "multitype",
                        "active": True,
                        "baseAbility": "multitype",
                        "condition": "402/402",
                        "details": "Arceus-Fighting",
                        "ident": "p2: Arceus",
                        "item": "fistplate",
                        "moves": [
                            "thunderbolt",
                            "shadowclaw",
                            "aerialace",
                            "judgment"
                        ],
                        "pokeball": "pokeball",
                        "stats": {
                            "atk": 245,
                            "def": 297,
                            "spa": 297,
                            "spd": 297,
                            "spe": 297
                        }
                    },
                    {
                        "ability": "sturdy",
                        "active": False,
                        "baseAbility": "sturdy",
                        "condition": "39/39",
                        "details": "Aggron, L10, M",
                        "ident": "p2: Aggron",
                        "item": "aggronite",
                        "moves": [
                            "roar",
                            "heavyslam",
                            "rockslide",
                            "toxic"
                        ],
                        "pokeball": "pokeball",
                        "stats": {
                            "atk": 32,
                            "def": 46,
                            "spa": 22,
                            "spd": 22,
                            "spe": 20
                        }
                    },
                    {
                        "ability": "multiscale",
                        "active": False,
                        "baseAbility": "multiscale",
                        "condition": "43/43",
                        "details": "Dragonite, L10, F",
                        "ident": "p2: Dragonite",
                        "item": "lumberry",
                        "moves": [
                            "dragonclaw",
                            "firepunch",
                            "roost",
                            "earthquake"
                        ],
                        "pokeball": "pokeball",
                        "stats": {
                            "atk": 37,
                            "def": 29,
                            "spa": 30,
                            "spd": 30,
                            "spe": 26
                        }
                    },
                    {
                        "ability": "levitate",
                        "active": False,
                        "baseAbility": "levitate",
                        "condition": "40/40",
                        "details": "Uxie, L10",
                        "ident": "p2: Uxie",
                        "item": "leftovers",
                        "moves": [
                            "psyshock",
                            "yawn",
                            "stealthrock",
                            "psychic"
                        ],
                        "pokeball": "pokeball",
                        "stats": {
                            "atk": 25,
                            "def": 36,
                            "spa": 25,
                            "spd": 36,
                            "spe": 29
                        }
                    },
                    {
                        "ability": "waterabsorb",
                        "active": False,
                        "baseAbility": "waterabsorb",
                        "condition": "39/39",
                        "details": "Cacturne, L10, F",
                        "ident": "p2: Cacturne",
                        "item": "leftovers",
                        "moves": [
                            "swordsdance",
                            "seedbomb",
                            "suckerpunch",
                            "spikes"
                        ],
                        "pokeball": "pokeball",
                        "stats": {
                            "atk": 33,
                            "def": 22,
                            "spa": 33,
                            "spd": 22,
                            "spe": 21
                        }
                    },
                    {
                        "ability": "sheerforce",
                        "active": False,
                        "baseAbility": "sheerforce",
                        "condition": "40/40",
                        "details": "Druddigon, L10, F",
                        "ident": "p2: Druddigon",
                        "item": "lifeorb",
                        "moves": [
                            "dragontail",
                            "suckerpunch",
                            "gunkshot",
                            "aerialace"
                        ],
                        "pokeball": "pokeball",
                        "stats": {
                            "atk": 34,
                            "def": 28,
                            "spa": 22,
                            "spd": 28,
                            "spe": 19
                        }
                    }
                ]
            }
        }        
        state_changes, indices = preprocess_observation_json(obs)

Reply:
Hmm, I'm thinking instead of env_pkmn's scrape_input method returning a string, it:
Checks for an error. If it fines one, crash
Checks for 'actionspace'. If it finds one, it updates its action space and drops the input
Checks for winning. If it does, it updates the reward accordingly
Checks for the dictionary, and if it finds it, json.loads it and returns the dictionary?
    Is this along the lines of what you are thinking?